### PR TITLE
Fix refcount tracking for CreateAndAttach

### PR DESCRIPTION
**Pull request by jzt**
Fixes a small bug in refcount tracking. Required for https://github.com/vmware/vic/pull/5581.

### CODE REVIEW

File: pkg/vsphere/disk/disk.go [hunk 1]
```diff
@@ -45,6 +45,40 @@ type Filesystem interface {
 	Unmount(path string) error
 }
 
+type ReferenceCount struct {
+	owner      string
+	refname    string
+	references int
```
Comment (vburenin) [inline pkg/vsphere/disk/disk.go]: references -> count

File: pkg/vsphere/disk/disk.go [hunk 2]
```diff
@@ -45,6 +45,40 @@ type Filesystem interface {
...
+
+func NewReferenceCount(o, n string) *ReferenceCount {
+	return &ReferenceCount{
+		owner:      o,
+		refname:    n,
+		references: 0,
+	}
+}
+
+func (r *ReferenceCount) Increment() {
```
Comment (vburenin) [inline pkg/vsphere/disk/disk.go]: I would consider using atomics in this case you would eliminate a whole class of a potential concurrency problems.

File: pkg/vsphere/disk/disk.go [hunk 3]
```diff
@@ -45,6 +45,40 @@ type Filesystem interface {
...
+	r.references++
+	log.Infof("Increased %s references for %s to %d", r.refname, r.owner, r.references)
```
Comment (vburenin) [inline pkg/vsphere/disk/disk.go]: This has to be a debug. However, I would even argue you need a log here at all. If you need to log a number of counts, I would add a String() methods for logging purposes somewhere else and I would log it where it is really necessary.

File: pkg/vsphere/disk/disk_manager.go [hunk 4]
```diff
@@ -142,6 +142,21 @@ func (m *Manager) toSpec(config *VirtualDiskConfig) *types.VirtualDisk {
...
 	defer trace.End(trace.Begin(config.DatastoreURI.String()))
 
+	// reuse the disk if it already exists and is attached
+	if _, err := findDiskByFilename(op, m.vm, config.DatastoreURI.String()); err == nil {
+		// if the disk is attached, it should be in the disk cache
+		d, err := m.Get(op, config)
+		if err != nil {
+			return nil, err
+		}
+		d.l.Lock()
```
Comment (vburenin) [inline pkg/vsphere/disk/disk_manager.go]: Using atomics inside ref counters would eliminate a need for these locks.

Comment (vburenin) [inline pkg/vsphere/disk/disk_manager.go]: Could you also rename "l" to "mu"? It's just inconsistent.

Comment (jzt) [inline pkg/vsphere/disk/disk_manager.go]: The locks aren't used specifically to bump reference counters, only in this case (if I recall correctly). They are used elsewhere while creating, attaching, etc.

File: pkg/vsphere/disk/disk.go (hunk 2 above)
Comment (jzt) [inline pkg/vsphere/disk/disk.go]: See below.. atomics are used in places where the disk needs to be locked, and these counters are wrapped in those locks.

File: pkg/vsphere/disk/disk.go (hunk 3 above)
Comment (jzt) [inline pkg/vsphere/disk/disk.go]: Removed

File: pkg/vsphere/disk/disk.go [hunk 5]
```diff
@@ -45,6 +45,43 @@ type Filesystem interface {
...
+	return &ReferenceCount{
+		owner:   o,
+		refname: n,
+		count:   0,
+	}
+}
+
+// Increment increases the reference count by one
+func (r *ReferenceCount) Increment() {
+	r.count++
```
Comment (chengwang86) [inline pkg/vsphere/disk/disk.go]: Do you think if we need to make these operations atomic by adding locks?

File: pkg/vsphere/disk/disk.go [hunk 6]
```diff
@@ -45,6 +45,43 @@ type Filesystem interface {
 	Unmount(path string) error
 }
 
+// ReferenceCount represents the number of references to a disk
+type ReferenceCount struct {
```
Comment (hmahmood) [inline pkg/vsphere/disk/disk.go]: I think just `Reference` seems closer to what this is actually.

File: pkg/vsphere/disk/disk.go (hunk 5 above)
Comment (hmahmood) [inline pkg/vsphere/disk/disk.go]: Use the functions in the `sync/atomic` package.

File: pkg/vsphere/disk/disk_manager_test.go [hunk 7]
```diff
@@ -158,6 +158,26 @@ func TestCreateAndDetach(t *testing.T) {
 		if !assert.NoError(t, cerr) {
 			return
 		}
+		refs := child.attachedRefs.Count()
```
Comment (hickeng) [inline pkg/vsphere/disk/disk_manager_test.go]: how about a separate unit test for the reference counting? Use goroutines to get some parallelism.

File: pkg/vsphere/disk/disk.go [hunk 8]
```diff
@@ -45,6 +45,43 @@ type Filesystem interface {
...
+func NewReferenceCount(o, n string) *ReferenceCount {
+	return &ReferenceCount{
+		owner:   o,
+		refname: n,
+		count:   0,
+	}
+}
+
+// Increment increases the reference count by one
+func (r *ReferenceCount) Increment() {
```
Comment (hickeng) [inline pkg/vsphere/disk/disk.go]: make this and decrement return the new value - it enables more types of usage.

File: pkg/vsphere/disk/disk.go [hunk 9]
```diff
@@ -45,6 +45,43 @@ type Filesystem interface {
...
+type ReferenceCount struct {
+	owner   string
+	refname string
+	count   int
+}
+
+// NewReferenceCount creates and returns a ReferenceCount initialized to 0
+func NewReferenceCount(o, n string) *ReferenceCount {
+	return &ReferenceCount{
+		owner:   o,
```
Comment (hickeng) [inline pkg/vsphere/disk/disk.go]: Owner is confusing when applied to locking structures - this is not the owner of a lock, it's the identifier of the resource.
Different name please.

File: pkg/vsphere/disk/disk.go (hunk 6 above)
Comment (hickeng) [inline pkg/vsphere/disk/disk.go]: This is an unbounded counting semaphore - can see why we're not implementing it using channels, but calling it a semaphore might be less confusing.

File: pkg/vsphere/disk/disk.go (hunk 5 above)
Comment (jzt) [inline pkg/vsphere/disk/disk.go]: There are already locks around this code albeit at a higher level. Are you sure you want to lock within a lock?

File: pkg/vsphere/disk/disk.go (hunk 6 above)
Comment (jzt) [inline pkg/vsphere/disk/disk.go]: Ehh... just `Semaphore`?  Feels weird, but I'll go with it for now unless someone has something more fitting. I agree, though, this is a count of references, not a reference.

File: pkg/vsphere/disk/disk.go (hunk 8 above)
Comment (jzt) [inline pkg/vsphere/disk/disk.go]: Fixed

File: pkg/vsphere/disk/disk.go (hunk 5 above)
Comment (jzt) [inline pkg/vsphere/disk/disk.go]: Fixed?

File: pkg/vsphere/disk/disk.go (hunk 6 above)
Comment (jzt) [inline pkg/vsphere/disk/disk.go]: Fixed

File: pkg/vsphere/disk/disk_manager.go (hunk 4 above)
Comment (jzt) [inline pkg/vsphere/disk/disk_manager.go]: Fixed?

File: pkg/vsphere/disk/disk.go (hunk 9 above)
Comment (jzt) [inline pkg/vsphere/disk/disk.go]: Renamed to `resource`.

File: pkg/vsphere/disk/disk_manager_test.go (hunk 7 above)
Comment (jzt) [inline pkg/vsphere/disk/disk_manager_test.go]: Added, let me know what you think.

**Pull request merged**